### ISSUE TITLE
Guide on how to use render props / custom components as arguments to a component

### ISSUE THREAD

**Issue by treshugart** [atlassian/tangerine#35]
## Overview

Components often require customization of the internal components it renders to change the look and feel - or even behavior - of it.

A good example of this would be an avatar that shows a default status icon. For the sake of this example, let's say we have `<Avatar />` and you want to be able to customize its `status.`

Due to the diffing algorithm that React and others use, what type of parameter we make `status` takes has implications on the lifecycle of nodes where the custom component is rendered. We should create guidelines around what type of parameter is passed, and for what use case.

## Proposal

When you take a custom component, there's two ways you can render it. The first form of customization is by taking a component `constructor`, `function` or `string` and rendering it as a JSX node:

```js
// avatar.js

import React from "react";

const DefaultStatus = props => <div {...props} />;

export default ({ status: Status = DefaultStatus }) => (
  <div>
    <Status>online</Status>
  </div>
);
```

The problem with this approach is that if you pass a SFC as `status`, it requires that it be the same instance as was previously rendered, else it thrashes the component lifecycle methods all components down the tree.

The second form of customization is using a render-prop style of prop, where you accept a function that takes arguments and returns a tree of nodes. The implementation of `Avatar` in this example only slightly changes, and due to the way `DefaultStatus` is implemented, it doesn't need to change.

```js
// avatar.js

import React from "react";

const DefaultStatus = props => <div {...props} />;

export default ({ status = DefaultStatus }) => (
  <div>
    {status({ children: 'online' })}
  </div>
);
```

You could, of course, change the signature of `DefaultStatus`, but by retaining the component-style `props` API, you make passing a SFC or `Function` interchangeable for the `status` prop. There are no known downsides to this approach other than possible ergonomic nits.

The fact the consumer would have to concern themselves with the issues that come with the first method to get performance by default, and distinguishing between an SFC and `Function` to also be able to use the second method, means that we should probably not consider the `<Status />` approach as an option, and instead follow the standard render-prop API.

## Any additional information

* Location: 'guides/components/components-vs-render-props.md`

**Comment by albertogasparin** [comment]
(Please forgive me if this is off topic, just trying to understand better where this proposal is going)

Note: I find the examples hard to understand, probably due to the way `DefaultStatus` is written. If I write `DefaultStatus` as a class, calling `{status({ children: 'online' })}` won't work. So we should not assume that the components we get as props are functional components, as people might change their implementation and then things suddenly break badly. 

So, if I understand this proposal correctly, we should require that every component passed in is wrapped in a function that creates the React element for it. Correct?
One downside of this technique is that it might move the burden of the optimisation up to the consumer. Let me explain with a really common example:

```js
import Avatar from 'avatar';
import MyStatusComponent from './my-status';

function MyApp () {
  return (
    <Avatar status={() => <MyStatusComponent />} />
  )
}
```

The big problem here is that to adapt to `Avatar` render prop API, I created a function that returns my component. What's happening however is that I'm creating a new function at every single render! Such inline anonymous function are often innocuous (eg in case of event callbacks) but in this case they might have serious performance consequences (plus it might be quite hard to spot and debug component suddenly unmounting and losing state).

So for internal use you might pay attention and avoid such scenario, but consumers might not pay the same level of attention (seen many many times in Jira). I wonder if the util `slot` could be smarter and (for instance) avoid the downsides of passing a component definition (eg. memoizing the react element?).

**Comment by theKashey** [comment]
It would be good to separate props, which accepts ReactNode, the ones which accept Components and the render prop ones by name.

```js
// you know that it is safe to render whatever you want here, _except_ passing real functional components, especaily with hooks.
<Avatar renderStatus={() => <MyStatusComponent />} />

// you know that it is safe to pass whatever you want here. Just a ReactNode.
<Avatar status={<MyStatusComponent />} />

// You know - your component shall preexists, or it could be a string - it does not matter
<Avatar statusComponent={MyStatusComponent} />
```

**Comment by treshugart** [comment]
> Note: I find the examples hard to understand, probably due to the way DefaultStatus is written. If I write DefaultStatus as a class, calling {status({ children: 'online' })} won't work. So we should not assume that the components we get as props are functional components, as people might change their implementation and then things suddenly break badly

Yeah, I forgot to change that to be clearly a function that returns React nodes (i.e. lower-casing the name). That said, it just so happens a functional component can have the same exact signature. WRT expectations, your type system will catch you if you try and call a class component. Apologies for the lack of clarity.

> So, if I understand this proposal correctly, we should require that every component passed in is wrapped in a function that creates the React element for it. Correct?
One downside of this technique is that it might move the burden of the optimisation up to the consumer. Let me explain with a really common example:

That’s a fair comment. It’s something I’ve definitely considered, but didn’t consider in the proposal. Happy to amend next week. This can be a problem in critical-path-components and is a well established caveat in other areas such as event handlers, as you noted. I think it’s a worthwhile trade-off for protecting from lifecycle thrashing, which generally has worse performance implications and is harder to prevent. At the very least, you can write lint rules for preventing it if it’s not already caught by linting rules that apply to binding event handlers.

Interesting thought regarding the slot element. I think it would be a good idea to explore this.
